internal/sftp/goclient_test.go

```diff
@@ -0,0 +1,288 @@
+	t.Helper()
+
+	var err error
+
+	srv := ssh.Server{
+		Addr: addr,
+		Handler: func(s ssh.Session) {
+			authorizedKey := gossh.MarshalAuthorizedKey(s.PublicKey())
+			io.WriteString(s, fmt.Sprintf("public key used by %s:\n", s.User()))
+			s.Write(authorizedKey)
+		},
+		PublicKeyHandler: pubkeyHandler,
+		SubsystemHandlers: map[string]ssh.SubsystemHandler{
+			"sftp": sftpHandler,
+		},
+	}
+
+	signer, err := hostKeySigner()
+	if err != nil {
+		t.Fatalf("SFTP server: %v", err)
+	}
+	srv.AddHostKey(signer)
+
+	errCh := make(chan error, 1)
+	go func() {
+		errCh <- srv.ListenAndServe()
+	}()
+
+	// Wait for the server to be ready
+	func() {
+		for {
+			select {
+			case err := <-errCh:
+				t.Fatalf("Couldn't start SFTP server: %v", err)
+			default:
+				conn, err := net.DialTimeout("tcp", addr, 1*time.Second)
+				if err == nil {
+					conn.Close()
+					return
+				}
+				time.Sleep(10 * time.Millisecond)
+			}
+		}
+	}()
+
+	t.Cleanup(func() { srv.Close() })
+	return &srv
+}
+
+func TestGoClient(t *testing.T) {
+	host, port, err := net.SplitHostPort(serverAddress)
+	if err != nil {
+		t.Fatalf("Bad serverAddress: %s", serverAddress)
+	}
+
+	_ = startSFTPServer(t, serverAddress)
+
+	// Start a listener on an open port and use the address to test a bad SFTP
+	// server address.
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatalf("couldn't start listener: %v", err)
+	}
+	defer listener.Close()
+	badHost, badPort, _ := net.SplitHostPort(listener.Addr().String())
+
+	type results struct {
+		Bytes int64
+		Paths []tfs.PathOp
+	}
+
+	type test struct {
+		name    string
+		cfg     sftp.Config
+		want    results
+		wantErr string
+	}
+	for _, tc := range []test{
+		{
+			name: "Uploads a file using a key with no passphrase",
+			cfg: sftp.Config{
+				Host:           host,
+				Port:           port,
+				KnownHostsFile: "./testdata/known_hosts",
+				PrivateKey: sftp.PrivateKey{
+					Path: "./testdata/clientkeys/test_ed25519",
+				},
+			},
+			want: results{
+				Bytes: 13,
+				Paths: []tfs.PathOp{tfs.WithFile("test.txt", "Testing 1-2-3")},
+			},
+		},
+		{
+			name: "Uploads a file using a key with a passphrase",
+			cfg: sftp.Config{
+				Host:           host,
+				Port:           port,
+				KnownHostsFile: "./testdata/known_hosts",
+				PrivateKey: sftp.PrivateKey{
+					Path:       "./testdata/clientkeys/test_pass_rsa",
+					Passphrase: "Backpack-Spirits6-Bronzing",
+				},
+			},
+			want: results{
+				Bytes: 13,
+				Paths: []tfs.PathOp{tfs.WithFile("test.txt", "Testing 1-2-3")},
+			},
+		},
+		{
+			name: "Errors when the key passphrase is wrong",
+			cfg: sftp.Config{
+				Host:           host,
+				Port:           port,
+				KnownHostsFile: "./testdata/known_hosts",
+				PrivateKey: sftp.PrivateKey{
+					Path:       "./testdata/clientkeys/test_pass_rsa",
+					Passphrase: "wrong",
+				},
+			},
+			wantErr: "SSH: couldn't parse private key with passphrase: x509: decryption password incorrect",
```

Comment:
Just sharing a thought for future consideration – when writing/wrapping error messages, I often find myself trying to decide whether to use phrases like "couldn't..." or "failed...", and I sometimes hesitate about whether to prefix the error with the method name or leave that context to the caller.

Anyways, I've found this today: https://github.com/uber-go/guide/blob/master/style.md#error-wrapping. It seems useful, or if we come up with our own rules I'd be happy to write them somewhere.

Reply:
I like the Uber style guide on error wrapping.  I'll try to follow it in the future. :+1: